components/series/stackedBar.js

```diff
@@ -6,77 +6,39 @@
         var stackedBar = function(selection) {
             var container;
 
-            // takes an object with values associated with each property, and
-            // converts it into an array of values. Each value has the xValue associated
-            // with it.
-            //
-            // For example, this object:
-            //
-            // obj = { county: 'North Tyneside', labour: 23, conservative: 55 }
-            //
-            // becomes this ...
-            //
-            // [
-            //   { xValue: 'North Tyneside', name: 'labour', previousValue: 0, currentValue: 23},
-            //   { xValue: 'North Tyneside', name: 'conservative', previousValue: 23, currentValue: 78},
-            // ]
-            function objectDatapointToArray(obj) {
-                var values = [];
-                var yTotal = 0;
-                var xVal = obj[stackedBar.xValueKey.value];
-                for (var propertyName in obj) {
-                    if (obj.hasOwnProperty(propertyName) && propertyName !== stackedBar.xValueKey.value) {
-                        var previous = yTotal;
-                        yTotal += Number(obj[propertyName]);
-                        values.push({
-                            'name': propertyName,
-                            'previousValue': previous,
-                            'currentValue': yTotal,
-                            'xValue': xVal
-                        });
-                    }
-                }
-                return values;
-            }
-
             selection.each(function(data) {
 
                 // add a 'root' g element on the first enter selection. This ensures
                 // that it is just added once
                 container = d3.select(this);
 
-                var keyFunction = function(d) {
-                    return d[stackedBar.xValueKey.value];
-                };
-                var g = fc.utilities.simpleDataJoin(container, 'stacked-bar', data, keyFunction);
+                // If the user has already set stacked data then we don't need to re-compute it.
```

Comment:
Be sure to consider how to manage cache invalidation of you do cache on the
element.

On Sun, 12 Apr 2015 23:47 James Leftley <notifications@github.com> wrote:

> In components/series/stackedBar.js
> <https://github.com/ScottLogic/d3-financial-components/pull/299#discussion_r28208437>
> :
>
> >              selection.each(function(data) {
> >
> >                  // add a 'root' g element on the first enter selection. This ensures
> >                  // that it is just added once
> >                  container = d3.select(this);
> >
> > -                var keyFunction = function(d) {
> > -                    return d[stackedBar.xValueKey.value];
> > -                };
> > -                var g = fc.utilities.simpleDataJoin(container, 'stacked-bar', data, keyFunction);
> > +                // If the user has already set stacked data then we don't need to re-compute it.
>
> FWIW - in my experience with this problem, caching directly on the element
> - instead of the component or selection - seemed to work relatively well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ScottLogic/d3-financial-components/pull/299/files#r28208437>
> .
>